src/numerics/SteadyStateSystem.cpp

```diff
@@ -0,0 +1,305 @@
+
+        // solve the transient problem
+        int status = newton().solve(x, r, *this, loglevel);
+
+        // successful time step. Copy the new solution in r to
+        // the current solution in x.
+        if (status >= 0) {
+            if (loglevel > 1) {
+                writelog("\nTimestep ({}) succeeded", n);
+            }
+            successiveFailures = 0;
+            m_nsteps++;
+            n += 1;
+            copy(r, r + m_size, x);
+            // No Jacobian evaluations were performed, so a larger timestep can be used
+            if (m_jac->nEvals() == j0) {
+                dt *= 1.5;
+            }
+            if (m_time_step_callback) {
+                m_time_step_callback->eval(dt);
+            }
+            dt = std::min(dt, m_tmax);
+            if (m_nsteps >= m_nsteps_max) {
+                throw CanteraError("OneDim::timeStep",
+                    "Took maximum number of timesteps allowed ({}) without "
+                    "reaching steady-state solution.", m_nsteps_max);
+            }
+        } else {
+            // No solution could be found with this time step.
+            // Decrease the stepsize and try again.
+            successiveFailures++;
+            if (loglevel == 1) {
+                writelog("\nTimestep failed");
+            } else if (loglevel > 1) {
+                writelog("\nTimestep ({}) failed", n);
+            }
+            if (successiveFailures > 2) {
+                debuglog("--> Resetting negative species concentrations", loglevel);
+                resetBadValues(x);
+                successiveFailures = 0;
+            } else {
+                debuglog("--> Reducing timestep", loglevel);
+                dt *= m_tfactor;
+                if (dt < m_tmin) {
+                    string err_msg = fmt::format(
+                        "Time integration failed. Minimum timestep ({}) reached.", m_tmin);
+                    throw CanteraError("OneDim::timeStep", err_msg);
+                }
+            }
+        }
+    }
+
+    // Write the final step to the log
+    if (loglevel == 1) {
+        double ss = ssnorm(x, r);
+        writelog("\n{:<5d}  {:<6.4e}   {:>7.4f}", n, dt, log10(ss));
+        writelog("\n============================");
+    } else if (loglevel > 1) {
+        double ss = ssnorm(x, r);
+        writelog("\nTimestep ({}) dt= {:<11.4e} log10(ss)= {:<7.4f}\n", n, dt, log10(ss));
+    }
+
+    // return the value of the last stepsize, which may be smaller
+    // than the initial stepsize
+    return dt;
+}
+
+double SteadyStateSystem::ssnorm(double* x, double* r)
+{
+    eval(x, r, 0.0, 0);
+    double ss = 0.0;
+    for (size_t i = 0; i < m_size; i++) {
+        ss = std::max(fabs(r[i]),ss);
+    }
+    return ss;
+}
+
+void SteadyStateSystem::setTimeStep(double stepsize, size_t n, const int* tsteps)
+{
+    m_tstep = stepsize;
+    m_steps.resize(n);
+    for (size_t i = 0; i < n; i++) {
+        m_steps[i] = tsteps[i];
+    }
```

Comment:
Is there a way to handle the case when `n != len(tsteps)`?

Reply:
No. One of the standard challenges of working with C arrays is that there is no inherent indication of length. Note that while the `SteadyStateSystem` class is new, this is just the implementation of what was `OneDim::setTimeStep`, so the signature needs to remain as-is.

Reply:
No. One of the standard challenges of working with C arrays is that there is no inherent indication of length. Note that while the `SteadyStateSystem` class is new, this is just the implementation of what was `OneDim::setTimeStep`, so the signature needs to remain as-is.

Reply:
Thanks. That's what I figured. Oh well! ☺️